c7n/query.py

```diff
@@ -418,8 +418,35 @@ def resources(self, query=None):
             query = {}
 
         resources = self.augment(self.source.resources(query))
+        resource_count = len(resources)
         self._cache.save(key, resources)
-        return self.filter_resources(resources)
+        resources = self.filter_resources(resources)
+
+        # Check if we're out of a policies execution limits.
+        self.check_resource_limit(len(resources), resource_count)
+        return resources
+
+    def check_resource_limit(self, selection_count, population_count):
+        """Check if policy's execution affects more resources then its limit.
+
+        Ideally this would be at a higher level but we've hidden
+        filtering behind the resource manager facade for default usage.
+        """
+        p = self.ctx.policy
+        if p.max_resources and p.max_resources > selection_count:
+            raise ResourceLimitExceeded(
+                ("policy: %s exceeding resource limit: {limit} "
+                 "found: {selection_count}") % p.name,
+                p.max_resources, selection_count, population_count)
+        elif p.max_resources_percent:
+            if (population_count * (
+                    p.max_resources_percent / 100.0) < selection_count):
+                raise ResourceLimitExceeded(
+                    ("policy: %s exceeding resource percent "
+                     "limit: {limit} found: {selection_count} "
```

Comment:
can you add a % after {limit}, the message is a bit unclear without it imo, on first read it appeared that the limit of resources was 10 resources, not 10%: 

```
c7n.exceptions.ResourceLimitExceeded: policy: c7n-rest-api-test exceeding resource percent limit: 10% found: 4 total: 4
```